docs/tutorials/demo_best_practices.py

```diff
@@ -0,0 +1,281 @@
+"""
+The cell below contains a helper function to obtain the data and code to obtain the default lakeFS credentials. It is otherwise not relevant to this demonstration.
+"""
+
+
+# %%
+def _maybe_urlretrieve(url: str, filename: str) -> str:
+    # Avoid API rate limit errors by downloading to a fixed local location
+    destination = Path(tempfile.gettempdir()) / "lakefs-spec-tutorials" / filename
+    destination.parent.mkdir(exist_ok=True, parents=True)
+    if destination.exists():
+        return str(destination)
+
+    outfile, _ = urllib.request.urlretrieve(url, str(destination))
+    return outfile
+
+
+outfile = _maybe_urlretrieve(
+    "https://archive-api.open-meteo.com/v1/archive?latitude=52.52&longitude=13.41&start_date=2010-01-01&end_date=2010-12-31&hourly=temperature_2m,relativehumidity_2m,rain,pressure_msl,surface_pressure,cloudcover,cloudcover_low,cloudcover_mid,cloudcover_high,windspeed_10m,windspeed_100m,winddirection_10m,winddirection_100m",
+    "weather-2010.json",
+)
+urllib.request.urlretrieve(
+    "https://raw.githubusercontent.com/aai-institute/lakefs-spec/main/docs/tutorials/.lakectl.yaml",
+    os.path.expanduser("~/.lakectl.yaml"),
+)
+# %% [markdown]
+"""
+## Define a data repository
+
+We got the data for the year 2010. That should be enough for initial prototyping.
+Later, however, we want to use more data. Since our dataset will be evolving, we implement data version control.
+This ensure the reproducibility of our experiments, enables collaboration with colleagues, and ensures our dynamic dataset to stay a valuable asset.
+
+To set up our versioning, we need to decide on a versioning tool, set up a repository, and define which data is considered in scope and should be versioned and which is not.
+
+Let us now set up a lakeFS repository using lakeFS-spec, our filesystem implementation for lakeFS.
+"""
+
+# %%
+fs = lakefs_spec.LakeFSFileSystem()
+
+REPO_NAME = "weatherpred"
+
+repo = lakefs.Repository(REPO_NAME, fs.client).create(storage_namespace=f"local://{REPO_NAME}")
+
+
+# %% [markdown]
+"""
+lakeFS works similarly to `git` as a versioning system.
+You can create *commits* that contain specific changes to the data.
+You can also work with *branches* to create an isolated view of the data.
+Every commit (on any branch) is identified by a commit SHA, a unique identifier obtained via a hashing function.
+
+## Branching Strategy
+
+We recommend following a branching strategy that ensures the data integrity on the main branch.
+Since we are about to do some data wrangling, we will branch off `main` and later merge back into it, once we are sure everything works as expected.
+"""
+
+# %%
+NEW_BRANCH_NAME = "transform-raw-data"
+
+with fs.transaction as tx:
+    fs.put(outfile, f"{REPO_NAME}/{NEW_BRANCH_NAME}/weather-2010.json")
+    tx.commit(repository=REPO_NAME, branch=NEW_BRANCH_NAME, message="Add 2010 weather data")
+
+# %% [markdown]
+"""
+## Utilize reusable and tested functions for data transformation
```

Comment:
Since we are communicating best practices, I would like to see type hints on all these functions, even though the make the signatures slightly longer.